File: programmers/etc/level3/이중우선순위큐.py
from heapq import *

def solution(operations):
    q, max_q = [], []
    for op in operations:
        cmd, num = op.split()
        num = int(num)
        if cmd == "I":
            heappush(q, num)
            heappush(max_q, -num)
        else:
            if q:
                if num == 1:
                    _max = heappop(max_q)
                    q.remove(-_max)
                else:
                    _min = heappop(q)
                    max_q.remove(-_min)                
                    
    return [-heappop(max_q), heappop(q)] if q else [0, 0]


operations = ["I 16", "I -5643", "D -1", "D 1", "D 1", "I 123", "D -1"]
print(solution(operations)) # [0, 0]

operations = ["I -45", "I 653", "D 1", "I -642", "I 45", "I 97", "D 1", "D -1", "I 333"]
print(solution(operations)) # [333, -45]

File: Java/programmers/etc/level3/이중우선순위큐.java
package CodeTest.Java.programmers.etc.level3;

import java.util.*;

public class 이중우선순위큐 {
    public int[] solution(String[] operations) {
        PriorityQueue<Integer> min_q = new PriorityQueue<>();
        PriorityQueue<Integer> max_q = new PriorityQueue<>(Collections.reverseOrder());
        
        for (String operation : operations) {
            String[] cmd = operation.split(" ");
            int num = Integer.parseInt(cmd[1]);
            
            if (cmd[0].equals("I")) {
                min_q.add(num);
                max_q.add(num);
            } else {
                if (!min_q.isEmpty()) {
                    if (num == 1) {
                        int _max = max_q.poll();
                        min_q.remove(_max);
                    } else {
                        int _min = min_q.poll();
                        max_q.remove(_min);
                    }
                }
            }
        }
        
        return min_q.isEmpty() ? new int[]{0, 0} : new int[]{max_q.peek(), min_q.peek()};
    }
}
